app/commands/user_track/generate_concepts_map_data.rb

```diff
@@ -0,0 +1,24 @@
+class UserTrack
+  class GenerateConceptsMapData
+    include Mandate
+
+    initialize_with :user_track
+
+    def call
+      Track::DetermineConceptsLayout.(user_track.track).tap { |data| data[:status] = concepts_status }
+    end
+
+    private
+    memoize
+    def concepts_status
+      all_concepts = user_track.track.concepts.map(&:slug)
+      available_concepts = user_track.available_concepts.map(&:slug)
+      learnt_concepts = user_track.learnt_concepts.map(&:slug)
+
+      {}.
+        tap { |h| all_concepts.each { |concept| h[concept] = :locked } }.
+        tap { |h| available_concepts.each { |concept| h[concept] = :unlocked } }.
+        tap { |h| learnt_concepts.each { |concept| h[concept] = :complete } }
+    end
```

Comment:
Nice. You could also DRY this up with a helper lambda.

```ruby
{}.tap do |output|
  set_status = ->(slugs, status) { slugs.each { |slug| output[slug] = status } }

  set_status.(all_concepts, :locked)
  set_status.(available_concepts, :unlocked)
  set_status.(learnt_concepts, :available)
end
```

Another suggestion, as this is a command, that you could pull out that lambda into a method, and then call the method each time, using an instance variable. As the aim of commands is that they're only called once, and can only be accessed through that command, having that DRY lambda as a helper function makes sense.

I'd merge this as it is though - I'm nitpicking and exploring code ideas only here 🙂